---
title: "MetaAnalysis_Plots"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

Meta-analysis was carried on three datasets Extend, EuGEI, and understanding society.METAL was used to run the meta-analysis. The script used for METAL is stored in */GWAS_bloodcelltypes/Scripts/METAL_script.txt*. The script used to clean the plink files after GWAS is stored in *GWAS_bloodcelltypes/Scripts/Adding_A2_to_plinkassoclinear_files.txt*

This report visualises the manhattan plot and qqplots of each cell type.

##CD8T

```{r cd8treadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
cd8t = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/Metal_Outputs/METAANALYSIS_IVWCD8T_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
cd8t_n <- cd8t
cd8t_n$ChrBp <- cd8t_n$MarkerName
cd8t_n <- separate(cd8t_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
cd8t_n$Chr <- as.numeric(as.character(cd8t_n$Chr))
cd8t_n$Bp <- as.numeric(as.character(cd8t_n$Bp))

colnames(cd8t_n)[colnames(cd8t_n) == 'MarkerName'] <- 'SNP'
colnames(cd8t_n)[colnames(cd8t_n) == 'Chr'] <- 'CHR'
colnames(cd8t_n)[colnames(cd8t_n) == 'P.value'] <- 'P'
colnames(cd8t_n)[colnames(cd8t_n) == 'Bp'] <- 'BP'
```


```{r Minimanhattan plot, include = FALSE, warning = F, message = F}
miniman <- function (data = NULL, chr = NULL, cpg = NULL, range = NULL, 
xlim = NULL, ylim = NULL, xlab = NULL, ylab = NULL, pch = 1, main = NULL, 
col = "black", cex = 1, result = NULL, pad = 30000, multiply = NULL, 
nullcol = "black", negcol = "black", poscol = "black", ESlevel = 0, ESdat = NULL, 
cexgene = 0.5, geneplot = TRUE, genelines = NULL, chrcol = NULL, 
mapcol = NULL, cpgcol = "forestgreen", cex.axis = 1, cex.lab = 1) {

# data should just be a dataframe
# chrcol and mapcol will automatically use 'CHR' and 'MAPINFO' but can be 
# replaced. chrcol column should just be numeric
# Can either provide a chr and range or cpg (must be rowname and have 
# mapinfo and chr in the file)
# result is the column who's p-values you want to plot
# pad is amount of padding to add ot the coordinates given
# cexgene is text size for labelling genes
# geneplot will be added by default but can be stopped if = "FALSE"
# genelines will autimatically be calculated by number of transcripts but 
# can be adjusted. If more transtcripts than genelines, several will be plotted 
# on the same line
# ESdat is the effect size column to colour points by
# ESlevel is target effect size to colour by
# nullcol is colour to plot if data doesn't meet ESlevel
# negcol is colour to plot if data meets -ESlevel 
# poscol is colour to plot if data meets +ESlevel 
# cpgcol is colour for cpg island track

require(TxDb.Hsapiens.UCSC.hg19.knownGene)
require(Homo.sapiens)
require(AnnotationHub)

if(missing(data))        {
stop("please provide data")
                }	

if(is.null(xlab)) {
	xlab = "Genomic Position"
	}

if(is.null(ylab)) {
	ylab = "-log10(p)"
	}	
	
if(is.null(main)) {
	main = ""
	}

if(is.null(chrcol)){
    chrcol="CHR"
    }
    
if(is.null(mapcol)){
    mapcol="MAPINFO"
    }    
      
if(missing(chr)) {
                        
if(missing(cpg)){

stop("please provide either 'chr' and 'range' or 'cpg'")
                
                }	
		          }

if(! is.null(cpg)){

	data[cpg,chrcol]->chr
	data[cpg,mapcol]->MI
	range = MI:MI
	
                  }	
	
if(is.null(xlim)) {
      xlim <- c((min(range) - pad),(max(range) + pad))
                  }

if(missing(result)){

stop("please provide result you would like to plot")
                
                   }	

par(xpd=TRUE)

data[which(data[,chrcol] == chr & data[,mapcol] >= min(range) - 
pad & data[,mapcol] <= max(range) + pad),]->y

if(is.null(ylim)) {
	ylim = c(min(-log10(y[,result])),max(-log10(y[,result])))
	}
	
if(is.null(chr)) {
	chr = y[,chrcol]
	}
	
if(is.null(cexgene)) {
	cexgene = 0.5
	}
	
if(is.null(geneplot)) {
	geneplot = TRUE
	}

ESlevel <- ESlevel/multiply

if(geneplot == "TRUE"){
layout(matrix(c(1,2), 2, 1, byrow = TRUE))
par(mar=c(3,3,1,2)+ 0.1, mgp=c(2,1,0)) 

if(!is.null(ESdat)){

plot(-log10(y[which(y[,ESdat] > -ESlevel & y[,ESdat] < ESlevel),result])~
y[which(y[,ESdat] > -ESlevel & y[,ESdat] < ESlevel),mapcol], xlim = xlim, 
ylim = ylim, cex = cex, pch = pch, xlab = xlab, ylab = ylab , xpd=FALSE, 
col=nullcol, cex.axis = cex.axis, cex.lab = cex.lab)
points(-log10(y[which(y[,ESdat] >= ESlevel),result])~
y[which(y[,ESdat] >= ESlevel),mapcol], col=poscol, cex = cex)
points(-log10(y[which(y[,ESdat] <= -ESlevel),result])~
y[which(y[,ESdat] <= -ESlevel),mapcol], col=negcol, cex = cex)
                   }

else{

plot(-log10(y[,result])~y[,mapcol], xlim = xlim, ylim = ylim, cex = cex, 
pch = pch, xlab = xlab, ylab = ylab , xpd=FALSE, col=nullcol)

     }

chr2=paste("chr", chr, sep="")

hub <- AnnotationHub()
query(hub, c("cpg","hg19"))
cpgs <- as.data.frame(hub[["AH5086"]])
cpgs[which(cpgs$seqnames == chr2 & cpgs$start >=min(range)-pad & 
cpgs$start <= max(range)+pad),]->cpg
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

gr <- GRanges(seqnames = chr2, ranges = IRanges(start = min(range)-pad, 
end = max(range)+pad))

subsetByOverlaps(transcripts(txdb), gr)->trans

res <- as.data.frame(transcripts(Homo.sapiens, columns=c("TXNAME","SYMBOL")))
as.data.frame(res[res$TXNAME %in% as.data.frame(trans)[,"tx_name"],"SYMBOL"])->m
colnames(m)<-"SYMBOL"
cbind(trans,m)->trans
trans[is.na(trans$SYMBOL) == "FALSE",]->trans

if(is.null(genelines)){
	genelines = nrow(trans)
	                  }

rep(1:max(genelines), times=ceiling(nrow(trans)/max(genelines)))->gl
gl[1:nrow(trans)]->gl
trans$gl <- gl
	
subsetByOverlaps(exonsBy(txdb, by = "tx"), gr)->exons
exons[rownames(trans)]->exons

par(mar=c(1,3,1,2)+ 0.1) 
plot(0,0,type="n", xlim=xlim, ylim=c(0,genelines +1), axes=FALSE, xlab="", 
ylab="", xpd=FALSE)

for(g in 1:nrow(cpg)){
polygon(c(cpg[g, "start"], cpg[g, "end"], cpg[g, "end"], cpg[g, "start"]), 
c((genelines+1)-0.2, (genelines+1)-0.2, (genelines+1)+0.2,(genelines+1)+0.2),
col=cpgcol, xpd=FALSE, border=cpgcol)
                     }

for(i in 1:nrow(trans)){

lines(c(as.data.frame(trans)[i,"start"], as.data.frame(trans)[i,"end"]), 
c(trans$gl[i],trans$gl[i]), xpd=FALSE)
text(as.data.frame(trans)[i,"start"]+((as.data.frame(trans)[i,"end"] - 
as.data.frame(trans)[i,"start"])/2) ,trans$gl[i] + 0.4 , trans$SYMBOL[i], 
cex=cexgene, xpd=FALSE)

for(x in 1:nrow(as.data.frame(exons[[i]]))){

polygon(c(as.data.frame(exons[[i]])[x,"start"], 
as.data.frame(exons[[i]])[x,"end"], as.data.frame(exons[[i]])[x,"end"], 
as.data.frame(exons[[i]])[x,"start"]), 
c(trans$gl[i]-0.2, trans$gl[i]-0.2, trans$gl[i]+0.2,trans$gl[i]+0.2),
col="black", xpd=FALSE)


                                           }
                      }
                      }
 
if(geneplot == "FALSE"){
plot.new()
if(!is.null(ESdat)){
par(mar=c(3,3,1,2)+ 0.1, mgp=c(2,1,0)) 

plot(-log10(y[which(y[,ESdat] > -ESlevel & y[,ESdat] < ESlevel),result])~
y[which(y[,ESdat] > -ESlevel & y[,ESdat] < ESlevel),mapcol], xlim = xlim, 
ylim = ylim, cex = cex, pch = pch, xlab = xlab, ylab = ylab , xpd=FALSE, 
col=nullcol, cex.axis = cex.axis, cex.lab = cex.lab)
points(-log10(y[which(y[,ESdat] >= ESlevel),result])~
y[which(y[,ESdat] >= ESlevel),mapcol], col=poscol, cex = cex)
points(-log10(y[which(y[,ESdat] <= -ESlevel),result])~
y[which(y[,ESdat] <= -ESlevel),mapcol], col=negcol, cex = cex)

                   }

else{

plot(-log10(y[,result])~y[,mapcol], xlim = xlim, ylim = ylim, cex = cex, 
pch = pch, xlab = xlab, ylab = ylab , xpd=FALSE, col=nullcol, cex.axis = cex.axis, 
cex.lab = cex.lab)

     }

                        }
                                                                }

```




```{r cd8tplot, warning = F, message = F}
pos = '+++'
neg = '---'

cd8t_n <- cd8t_n[which(cd8t_n$Direction %in% c(pos,neg)),]

bonfP<-0.05/nrow(cd8t_n)
par(mfrow = c(1,2))

manhattan(cd8t_n, genomewide = -log10(bonfP), main = "CD8T",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(cd8t_n$P, main = "CD8T")

```

We can zoom in to see the most significant probe.

```{r cd8t minimanhattan, warning = F, message = F}

cd8t_n <- cd8t_n[order(cd8t_n$P),]
sig_cd8t <- cd8t_n[1,]
manhattan(subset(cd8t_n, CHR == sig_cd8t$CHR), highlight = cd8t_n[cd8t_n$P >= bonfP, 'SNP'],suggestiveline = -log10(5e-5), genomewide = -log10(bonfP), xlim = c((sig_cd8t$BP - 10000), (sig_cd8t$BP + 10000)), main = sig_cd8t$CHR)

```

What is the PM plot like for this cell type variant?

```{r cd8t annotation, warning = F, message = F}

```

##CD4T

```{r CD4Treadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
CD4T = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWCD4T_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
CD4T_n <- CD4T[which(CD4T$Direction %in% c(pos,neg)),]
CD4T_n$ChrBp <- CD4T_n$MarkerName
CD4T_n <- separate(CD4T_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
CD4T_n$Chr <- as.numeric(as.character(CD4T_n$Chr))
CD4T_n$Bp <- as.numeric(as.character(CD4T_n$Bp))

colnames(CD4T_n)[colnames(CD4T_n) == 'MarkerName'] <- 'SNP'
colnames(CD4T_n)[colnames(CD4T_n) == 'Chr'] <- 'CHR'
colnames(CD4T_n)[colnames(CD4T_n) == 'P.value'] <- 'P'
colnames(CD4T_n)[colnames(CD4T_n) == 'Bp'] <- 'BP'
```

```{r CD4Tplots, warning = F, message = F}
bonfP<-0.05/nrow(CD4T_n)
par(mfrow = c(1,2))

manhattan(CD4T_n, genomewide = -log10(bonfP), main = "CD4T",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(CD4T_n$P, main = "CD4T")

```
##Bcell

```{r Bcellreadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
Bcell = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWBcell_1.txt",stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
Bcell_n <- Bcell[which(Bcell$Direction %in% c(pos,neg)),]
Bcell_n$ChrBp <- Bcell_n$MarkerName
Bcell_n <- separate(Bcell_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
Bcell_n$Chr <- as.numeric(as.character(Bcell_n$Chr))
Bcell_n$Bp <- as.numeric(as.character(Bcell_n$Bp))

colnames(Bcell_n)[colnames(Bcell_n) == 'MarkerName'] <- 'SNP'
colnames(Bcell_n)[colnames(Bcell_n) == 'Chr'] <- 'CHR'
colnames(Bcell_n)[colnames(Bcell_n) == 'P.value'] <- 'P'
colnames(Bcell_n)[colnames(Bcell_n) == 'Bp'] <- 'BP'
```

```{r Bcellplot, warning = F, message = F}
bonfP<-0.05/nrow(Bcell_n)
par(mfrow = c(1,2))

manhattan(Bcell_n, genomewide = -log10(bonfP), main = "Bcell",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(Bcell_n$P, main = "Bcell")

```
##NK

```{r NKreadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
NK = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWNK_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
NK_n <- NK[which(NK$Direction %in% c(pos,neg)),]
NK_n$ChrBp <- NK_n$MarkerName
NK_n <- separate(NK_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
NK_n$Chr <- as.numeric(as.character(NK_n$Chr))
NK_n$Bp <- as.numeric(as.character(NK_n$Bp))

colnames(NK_n)[colnames(NK_n) == 'MarkerName'] <- 'SNP'
colnames(NK_n)[colnames(NK_n) == 'Chr'] <- 'CHR'
colnames(NK_n)[colnames(NK_n) == 'P.value'] <- 'P'
colnames(NK_n)[colnames(NK_n) == 'Bp'] <- 'BP'
```

```{r NKplot, warning = F, message = F}
bonfP<-0.05/nrow(NK_n)
par(mfrow = c(1,2))

manhattan(NK_n, genomewide = -log10(bonfP), main = "NK",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(NK_n$P, main = "NK")

```
##Neu

```{r Neureadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
Neu <- read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWNeu_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
Neu_n <- Neu[which(Neu$Direction %in% c(pos,neg)),]
Neu_n$ChrBp <- Neu_n$MarkerName
Neu_n <- separate(Neu_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
Neu_n$Chr <- as.numeric(as.character(Neu_n$Chr))
Neu_n$Bp <- as.numeric(as.character(Neu_n$Bp))

colnames(Neu_n)[colnames(Neu_n) == 'MarkerName'] <- 'SNP'
colnames(Neu_n)[colnames(Neu_n) == 'Chr'] <- 'CHR'
colnames(Neu_n)[colnames(Neu_n) == 'P.value'] <- 'P'
colnames(Neu_n)[colnames(Neu_n) == 'Bp'] <- 'BP'
```

```{r Neuplot, warning = F, message = F}
bonfP<-0.05/nrow(Neu_n)
par(mfrow = c(1,2))

manhattan(Neu_n, genomewide = -log10(bonfP), main = "Neu",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(Neu_n$P, main = "Neu")

```


##Gran

```{r Granreadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
Gran = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWGran_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
Gran_n <- Gran[which(Gran$Direction %in% c(pos,neg)),]
Gran_n$ChrBp <- Gran_n$MarkerName
Gran_n <- separate(Gran_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
Gran_n$Chr <- as.numeric(as.character(Gran_n$Chr))
Gran_n$Bp <- as.numeric(as.character(Gran_n$Bp))

colnames(Gran_n)[colnames(Gran_n) == 'MarkerName'] <- 'SNP'
colnames(Gran_n)[colnames(Gran_n) == 'Chr'] <- 'CHR'
colnames(Gran_n)[colnames(Gran_n) == 'P.value'] <- 'P'
colnames(Gran_n)[colnames(Gran_n) == 'Bp'] <- 'BP'
```

```{r Granplot, warning = F, message = F}
bonfP<-0.05/nrow(Gran_n)
par(mfrow = c(1,2))

manhattan(Gran_n, genomewide = -log10(bonfP), main = "Gran",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(Gran_n$P, main = "Gran")

```


##Eos

```{r Eosreadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
Eos = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWEos_1.txt", stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
Eos_n <- Eos[which(Eos$Direction %in% c(pos,neg)),]
Eos_n$ChrBp <- Eos_n$MarkerName
Eos_n <- separate(Eos_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
Eos_n$Chr <- as.numeric(as.character(Eos_n$Chr))
Eos_n$Bp <- as.numeric(as.character(Eos_n$Bp))

colnames(Eos_n)[colnames(Eos_n) == 'MarkerName'] <- 'SNP'
colnames(Eos_n)[colnames(Eos_n) == 'Chr'] <- 'CHR'
colnames(Eos_n)[colnames(Eos_n) == 'P.value'] <- 'P'
colnames(Eos_n)[colnames(Eos_n) == 'Bp'] <- 'BP'
```

```{r Eosplot, warning = F, message = F}
bonfP<-0.05/nrow(Eos_n)
par(mfrow = c(1,2))

manhattan(Eos_n, genomewide = -log10(bonfP), main = "Eos",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow") )

qq(Eos_n$P, main = "Eos")

```


##Mono

```{r Monoreadin, include = FALSE, message = F, warning = F}
pos = '+++'
neg = '---'
Mono = read.table("/mnt/data1/GWAS_bloodcelltypes/MetaAnalysis_IVW/METAANALYSIS_IVWMono_1.txt",stringsAsFactors = F, header = T)

library(qqman)
library(tidyr)
Mono_n <- Mono[which(Mono$Direction %in% c(pos,neg)),]
Mono_n$ChrBp <- Mono_n$MarkerName
Mono_n <- separate(Mono_n, ChrBp, into= c("Chr", "Bp"), sep = ":")
Mono_n$Chr <- as.numeric(as.character(Mono_n$Chr))
Mono_n$Bp <- as.numeric(as.character(Mono_n$Bp))

colnames(Mono_n)[colnames(Mono_n) == 'MarkerName'] <- 'SNP'
colnames(Mono_n)[colnames(Mono_n) == 'Chr'] <- 'CHR'
colnames(Mono_n)[colnames(Mono_n) == 'P.value'] <- 'P'
colnames(Mono_n)[colnames(Mono_n) == 'Bp'] <- 'BP'
```

```{r Monoplot, warning = F, message = F}
bonfP<-0.05/nrow(Mono_n)

par(mfrow = c(1,2))
manhattan(Mono_n, genomewide = -log10(bonfP), main = "Mono",
          suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow"))

qq(Mono_n$P, main = "Mono")

```
