##### Title: QC BDR Imputed 
##### Date: 13-03-2019
##### Author: Gemma Shireby
##### email: gs470@exeter.ac.uk


# Genotyping QC PIPELINE ADAPTED FROM: https://github.com/MareesAT/GWA_tutorial/
# For a thorough theoretical explanation of all QC steps look at article: â€œA tutorial on conducting Genome-Wide-Association Studies: Quality control and statistical analysisâ€ (https://www.ncbi.nlm.nih.gov/pubmed/29484742).


##############################################################
##############################################################
##############################################################

## POST IMPUTATION - ONLY KEEP SNPS WHICH REACH INFO VALUE OF 0.3

FILE PATH: /mnt/data1/BDR/Genetic/Imputed/BDR_post_impute_1-22_plink2_R20.3
cd  /mnt/data1/BDR/Genetic/Imputed/
### Step 1 ### 

# Investigate missingness per individual and per SNP and make histograms.
plink2 --bfile /mnt/data1/BDR/Genetic/Imputed/BDR_post_impute_1-22_plink_R20.3 --missing    
# output: plink2.imiss and plink2.lmiss, these files show respectively the proportion of missing SNPs per individual and the proportion of missing individuals per SNP.

# no samples fail for missingness

# Generate plots to visualize the missingness results.
Rscript --vanilla hist_miss.R

# Delete SNPs and individuals with high levels of missingness, explanation of this and all following steps can be found in box 1 and table 1 of the article mentioned in the comments of this script.
# It is good practice to start the QC with these non-stringent thresholds.  
# Delete SNPs with missingness >0.1
plink2 --bfile /mnt/data1/BDR/Genetic/Imputed/Batch1_2_imputed/BDR_post_impute_1-22_plink_R20.3 --geno 0.1 --mind 0.1 --make-bed --out BDR_r3_2

## you can start with less stringent thresholds like above and then use more stringent thresholds if you think this would suit your data - in this case we kept the same threshold but this is why there are two steps 

# Delete individuals with missingness >0.1
plink2 --bfile BDR_r3_2 --mind 0.1 --make-bed --out BDR_r3_3

# Delete SNPs with missingness >0.1.
plink2 --bfile BDR_r3_3 --geno 0.1 --make-bed --out BDR_r3_4


## no samples removed for missingness - either individual or snp

################################################################### AS WE HAVE ONLY IMPUTE CHR 1-22 WE CANNOT DO SEX CHECK - I'VE LEFT THESE STEPS IN BUT BLOCKED OUT FOR FUTURE REF
# ### Step2 #### 

# # Check for sex discrepancy.
# # Subjects who were a priori determined as females must have a F value of <0.2, and subjects who were a priori determined as males must have a F value >0.8. This F value is based on the X chromosome inbreeding (homozygosity) estimate.
# # Subjects who do not fulfil these requirements are flagged "PROBLEM" by PLINK.

# plink2 --bfile BDR_r3_5 --check-sex 

# # Generate plots to visualize the sex-check results.
# Rscript --vanilla gender_check.R
# # These checks indicate that there is one woman with a sex discrepancy, F value of 0.99. (When using other datasets often a few discrepancies will be found). 

# # The following two scripts can be used to deal with individuals with a sex discrepancy.
# # Note, please use one of the two options below to generate the bfile hapmap_r23a_6, this file we will use in the next step of this tutorial.

# # 1) Delete individuals with sex discrepancy.
# grep "PROBLEM" plink2.sexcheck| awk '{print$1,$2}'> sex_discrepancy.txt
# # This command generates a list of individuals with the status
# plink2 --bfile BDR_r3_5 --remove sex_discrepancy.txt --make-bed --out BDR_r3_6 


# # 2) impute-sex.
# #plink2 --bfile BDR_r3_5 --impute-sex --make-bed --out BDR_r3_6
# # This imputes the sex based on the genotype information into your data set.

###################################################
### Step 3 ### 

# Generate a bfile with autosomal SNPs only and delete SNPs with a low minor allele frequency (MAF) - should only be autosomal as this was done pre-imputation 

# Select autosomal SNPs only (i.e., from chromosomes 1 to 22).
awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' BDR_r3_5.bim > snp_1_22.txt
plink2 --bfile BDR_r3_5 --extract snp_1_22.txt --make-bed --out BDR_r3_7

# Generate a plot of the MAF distribution.
plink2 --bfile BDR_r3_2 --freq --out MAF_check
Rscript --vanilla MAF_check.R

# Remove SNPs with a low MAF frequency.
plink2 --bfile BDR_r3_2 --maf 0.01 --make-bed --out BDR_r3_8
# 6550242  SNPs are left
# A conventional MAF threshold for a regular GWAS is between 0.01 or 0.05, depending on sample size.


####################################################
### Step 4 ###

# Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).
# Check the distribution of HWE p-values of all SNPs.

plink2 --bfile BDR_r3_8 --hardy
# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 
awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plink2zoomhwe.hwe
Rscript --vanilla hwe.R

## 0 samples fail 


# By default the --hwe option in plink2 only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.
plink2 --bfile BDR_r3_8 --hwe 0.0001 --make-bed --out HapMap_hwe_filter_step1


# --hwe: 481 variants removed due to Hardy-Weinberg exact test.
# 2583966 variants and 819 people pass filters and QC.



## 

# The HWE threshold for the cases filters out only SNPs which deviate extremely from HWE. 
# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed


# Theoretical background for this step is given in our accompanying article: https://www.ncbi.nlm.nih.gov/pubmed/29484742 .

############################################################
### step 5 ###

# Generate a plot of the distribution of the heterozygosity rate of your subjects.
# And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean.

# Checks for heterozygosity are performed on a set of SNPs which are not highly correlated.
# Therefore, to generate a list of non-(highly)correlated SNPs, we exclude high inversion regions (inversion.txt [High LD regions]) and prune the SNPs using the command --indep-pairwiseâ€™.
# The parameters â€˜50 5 0.2â€™ stand respectively for: the window size, the number of SNPs to shift the window at each step, and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously.

plink2 --bfile BDR_r3_8 --exclude inversion.txt --range --indep-pairwise 50 5 0.2 --out indepSNP
# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the tutorial.

plink2 --bfile BDR_r3_8 --extract indepSNP.prune.in --het --out R_check
# This file contains your pruned data set.

# Plot of the heterozygosity rate distribution
Rscript --vanilla check_heterozygosity_rate.R

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
# For data manipulation we recommend using UNIX. However, when performing statistical calculations R might be more convenient, hence the use of the Rscript for this step:
Rscript --vanilla heterozygosity_outliers_list.R # n = 7

# Output of the command above: fail-het-qc.txt .
# When using our example data/the HapMap data this list contains 2 individuals (i.e., two individuals have a heterozygosity rate deviating more than 3 SD's from the mean).
# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

## contains 7 samples

# Remove heterozygosity rate outliers.
plink2 --bfile BDR_r3_8 --remove het_fail_ind.txt --make-bed --out BDR_r3_10


############################################################
### step 6 ###

# It is essential to check datasets you analyse for cryptic relatedness.
# Assuming a random population sample we are going to exclude all individuals above the pihat threshold of 0.2 in this tutorial.

# Check for relationships between individuals with a pihat > 0.2.
plink2 --bfile BDR_r3_10 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2 ## 0 related samples 


### remove duplicated SNPS

plink2 --bfile /mnt/data1/BDR/Genetic/Imputed/Batch1_2_imputed/BDR_batch1_and2_imputed_QCd --make-bed --out BDR_both_batches_imputed_QCd

plink2 --bfile BDR_both_batches_imputed_QCd --update-ids /mnt/data1/BDR/Genetic/Imputed/BDR_geno_IDs.txt --make-bed --out BDR_both_batches_imputed_QCd_updatedIDs
################################################################################################################################

# FINISHED BDR IMPUTED QC: /mnt/data1/BDR/Genetic/Imputed/BDR_imputed_QCd
