## Script for QC post Imputation of genotype data

###############			Step 1			###############
#Check missing genotypes for individuals AND snps

# Investigate missingness per individual and per SNP and make histograms.
plink2 --bfile /mnt/data1/GWAS_bloodcelltypes/Understanding_Society/data_filtered_1_sex_upd --missing    

Rscript --vanilla hist_miss.R
# output: plink2.imiss and plink2.lmiss, these files show respectively the proportion of missing SNPs per individual and the proportion of missing individuals per SNP.
## RESULTS - All Zero


# Delete SNPs and individuals with high levels of missingness, explanation of this and all following steps can be found in box 1 and table 1 of the article mentioned in the comments of this script.
# It is good practice to start the QC with these non-stringent thresholds.  
# Delete SNPs with missingness >0.1
plink2 --bfile /mnt/data1/GWAS_bloodcelltypes/Understanding_Society/data_filtered_1_sex_upd --geno 0.1 --mind 0.1 --make-bed --out data_filtered_1_sex_upd_2

wc -l /mnt/data1/GWAS_bloodcelltypes/Understanding_Society/data_filtered_1_sex_upd.fam
wc -l /mnt/data1/GWAS_bloodcelltypes/Understanding_Society/data_filtered_1_sex_upd.bim
wc -l data_filtered_1_sex_upd_2.fam
wc -l data_filtered_1_sex_upd_2.bim

## RESULTS - No indivudals or snps were removed
## –- min 0.1   : Before 1111 samples. After  1111 samples
## -- geno 0.1 : Before 6513638 SNPs. After 6513638 SNPs 



###############			Step 2			###############
#check incorrect sex 



###############			Step 3			###############
#Check Minor Allele Frequency (typically do not want anything that is less than 5% (0.05))
# Generate a bfile with autosomal SNPs only and delete SNPs with a low minor allele frequency (MAF) - should only be autosomal as this was done pre-imputation 
# Select autosomal SNPs only (i.e., from chromosomes 1 to 22).
awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' data_filtered_1_sex_upd_2.bim > snp_1_22.txt
plink2 --bfile data_filtered_1_sex_upd_2 --extract snp_1_22.txt --make-bed --out data_filtered_1_sex_upd_3

# Generate a plot of the MAF distribution.
plink2 --bfile data_filtered_1_sex_upd_3 --freq --out MAF_check
Rscript --vanilla MAF_check.R

# Remove SNPs with a low MAF frequency.
plink2 --bfile data_filtered_1_sex_upd_3 --maf 0.01 --make-bed --out data_filtered_1_sex_upd_4
wc -l data_filtered_1_sex_upd_3.bim
wc -l data_filtered_1_sex_upd_4.bim 

# -- maf 0.01 : Before  6513638 SNPs. After 6466597 SNPs (47,041 snps removed)
# Select this MAF at 0.01 at the moment as we have a small sample size for this GWAS study.

###############			Step 4			###############

# Delete SNPs which are not in Hardy-Weinberg equilibrium (HWE).
# Check the distribution of HWE p-values of all SNPs.

plink2 --bfile data_filtered_1_sex_upd_4 --hardy
# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 
awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plinkzoomhwe.hwe
Rscript --no-save hwe.R

# By default the --hwe option in plink only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls, followed by a less stringent threshold for the case data.
plink2 --bfile data_filtered_1_sex_upd_4 --hwe 0.0001 --make-bed --out data_filtered_1_sex_upd_4_hwe
wc -l data_filtered_1_sex_upd_4_hwe.bim 

## -- hwe 0.0001 : Before 6466597 SNPs. After 6466368 SNPs (229 snps removed)

# Theoretical background for this step is given in our accompanying article: https://www.ncbi.nlm.nih.gov/pubmed/29484742 .

###############			Step 5			###############
# Heterozygosity 
# Generate a plot of the distribution of the heterozygosity rate of your subjects.
# And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean.

# Checks for heterozygosity are performed on a set of SNPs which are not highly correlated.
# Therefore, to generate a list of non-(highly)correlated SNPs, we exclude high inversion regions (inversion.txt [High LD regions]) and prune the SNPs using the command --indep-pairwiseâ€™.
# The parameters â€˜50 5 0.2â€™ stand respectively for: the window size, the number of SNPs to shift the window at each step, and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously.

plink2 --bfile data_filtered_1_sex_upd_4_hwe --exclude inversion.txt --range --indep-pairwise 50 5 0.2 --out indepSNP
# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the tutorial.

plink2 --bfile data_filtered_1_sex_upd_4_hwe  --extract indepSNP.prune.in --het --out R_check
# This file contains your pruned data set.

# Plot of the heterozygosity rate distribution
Rscript --vanilla check_heterozygosity_rate.R

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
# For data manipulation we recommend using UNIX. However, when performing statistical calculations R might be more convenient, hence the use of the Rscript for this step:
Rscript --vanilla heterozygosity_outliers_list.R # n = 2

# Output of the command above: fail-het-qc.txt .
# When using our example data/the HapMap data this list contains 2 individuals (i.e., two individuals have a heterozygosity rate deviating more than 3 SD's from the mean).
# Adapt this file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers.
plink2 --bfile data_filtered_1_sex_upd_4_hwe --remove het_fail_ind.txt --make-bed --out data_filtered_1_sex_upd_5

wc -l data_filtered_1_sex_upd_4_hwe.fam
wc -l data_filtered_1_sex_upd_5.fam


###############			Step 6			###############
# Relatedness 

# It is essential to check datasets you analyse for cryptic relatedness.
# Assuming a random population sample we are going to exclude all individuals above the pihat threshold of 0.2 in this tutorial.

# Check for relationships between individuals with a pihat > 0.2.
plink2 --bfile data_filtered_1_sex_upd_5 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2 

#56 samples appear to have a IBD score of 0.2 and higher (several samples at 0.5 threshold). Remove these related samples

# For each pair of 'related' individuals with a pihat > 0.2, we recommend to remove the individual with the lowest call rate. 
plink2 --bfile data_filtered_1_sex_upd_5 --missing

# Delete the individuals IBD greater than 0.2
plink2 --bfile data_filtered_1_sex_upd_5  --remove plink.nosex --make-bed --out data_filtered_1_sex_upd_6

wc -l data_filtered_1_sex_upd_5.fam
wc -l data_filtered_1_sex_upd_6.fam

################################################################################################################################

